bugsnag-android-core/src/main/java/com/bugsnag/android/DeviceIdStore.kt

```diff
@@ -0,0 +1,179 @@
+            }
+        } catch (exc: IOException) {
+            logger.w("Failed to created device ID file", exc)
+        }
+        this.synchronizedStreamableStore = SynchronizedStreamableStore(file)
+    }
+
+    /**
+     * Loads the device ID from its file system location. Device IDs are UUIDs which are
+     * persisted on a per-install basis. This method is thread-safe and multi-process safe.
+     *
+     * If no device ID exists then the legacy value stored in [SharedPreferences] will
+     * be used. If no value is present then a random UUID will be generated and persisted.
+     */
+    fun loadDeviceId(context: Context): String? {
+        val sharedPrefMigrator = SharedPrefMigrator(context)
+        val legacyDeviceId = sharedPrefMigrator.loadDeviceId()
+
+        return loadDeviceId(logger) {
+            when (legacyDeviceId) {
+                null -> UUID.randomUUID()
+                else -> UUID.fromString(legacyDeviceId)
+            }
+        }
+    }
+
+    internal fun loadDeviceId(logger: Logger, uuidProvider: () -> UUID): String? {
+        return try {
+            // optimistically read device ID without a lock - the majority of the time
+            // the device ID will already be present so no synchronization is required.
+            val deviceId = loadDeviceIdInternal(logger)
+
+            if (deviceId?.id != null) {
+                deviceId.id
+            } else {
+                return persistNewDeviceUuid(logger, uuidProvider)
+            }
+        } catch (exc: Throwable) {
+            logger.w("Failed to load device ID", exc)
+            null
+        }
+    }
+
+    /**
+     * Loads the device ID from the file.
+     *
+     * If the file has zero length it can't contain device ID, so reading will be skipped.
+     */
+    private fun loadDeviceIdInternal(logger: Logger): DeviceId? {
+        if (file.length() > 0) {
+            try {
+                return synchronizedStreamableStore.load(DeviceId.Companion::fromReader)
+            } catch (exc: IOException) {
+                logger.w("Failed to load device ID", exc)
+            }
+        }
+        return null
+    }
+
+    /**
+     * Write a new Device ID to the file.
+     */
+    private fun persistNewDeviceUuid(
+        logger: Logger,
+        uuidProvider: () -> UUID
+    ): String? {
+        return try {
+            // acquire a FileLock to prevent Clients in different processes writing
+            // to the same file concurrently
+            file.outputStream().channel.use { channel ->
+                persistNewDeviceIdWithLock(channel, logger, uuidProvider)
+            }
+        } catch (exc: Throwable) {
+            logger.w("Failed to persist device ID", exc)
+            null
+        }
+    }
+
+    private fun persistNewDeviceIdWithLock(
+        channel: FileChannel,
+        logger: Logger,
+        uuidProvider: () -> UUID
+    ): String? {
+        val lock = waitForFileLock(channel) ?: return null
```

Comment:
What lock is the `synchronizedStreamableStore` providing on top of this? I thought that was the purpose of the generic store?

If it's because we need to check the device ID once inside the lock, perhaps the `synchronizedStreamableStore` could be extended to allow this lock to be taken by a client ahead of persisting?

Reply:
There are two mechanisms in use: a [ReentrantReadWriteLock](https://developer.android.com/reference/java/util/concurrent/locks/ReentrantReadWriteLock) which is used by `SynchronizedStreamableStore` to synchronize across multiple threads, and a [FileLock](https://developer.android.com/reference/java/nio/channels/FileLock) which is used by `DeviceIdStore` to synchronize across multiple processes.

`SynchronizedStreamableStore` is effectively the replacement for `SharedPreferences`. When saving/loading the user information it isn't necessary to be multi process safe as each process will store this information in a separate file. Therefore as the device ID file is the only one that can be accessed across multiple processes, I've used the file locking in the `DeviceIdStore` only.

Reply:
OK - yes, I see. Noted that you've said this in the PR description, but can you add the above explanation to the method or class comments so it's obvious for future readers.

Reply:
Updated docs to include some more information on this